pkg/healthchecker/health_checker.go

```diff
@@ -0,0 +1,147 @@
+/*
+Copyright 2020 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package healthchecker
+
+import (
+	"fmt"
+	"net/http"
+	"os/exec"
+	"time"
+
+	"github.com/golang/glog"
+
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/node-problem-detector/cmd/healthchecker/options"
+	"k8s.io/node-problem-detector/pkg/healthchecker/types"
+)
+
+type healthChecker struct {
+	enableRepair    bool
+	healthCheckFunc func() bool
+	// The repair is "best-effort" and ignores the error from the underlying actions.
+	// The bash commands to kill the process will fail if the service is down and hence ignore.
+	repairFunc         func()
+	crictlPath         string
+	healthCheckTimeout time.Duration
+	coolDownTime       time.Duration
+}
+
+// NewHealthChecker returns a new health checker configured with the given options.
+func NewHealthChecker(hco *options.HealthCheckerOptions) (types.HealthChecker, error) {
+	hc := &healthChecker{
+		enableRepair:       hco.EnableRepair,
+		crictlPath:         hco.CriCtlPath,
+		healthCheckTimeout: hco.HealthCheckTimeout,
+		coolDownTime:       hco.CoolDownTime,
+	}
+	switch hco.Component {
+	case types.KubeletComponent:
+		hc.healthCheckFunc = kubeletHealthCheck
+		hc.repairFunc = kubeletRepair
+	case types.ContainerRuntimeComponent:
+		hc.healthCheckFunc = getContainerRuntimeHealthCheck(hco.ContainerRuntime, hco.CriCtlPath)
+		hc.repairFunc = getContainerRuntimeRepair(hco.ContainerRuntime)
+	default:
+		return nil, fmt.Errorf("invalid component : %+v specified in health check options", hco.Component)
+	}
+	return hc, nil
+}
+
+// CheckHealth checks for the health of the component and tries to repair if enabled.
+// Returns true if healthy, false otherwise.
+func (hc *healthChecker) CheckHealth() bool {
+	// Poll till the health check timeout for the component to be up.
+	if err := wait.PollImmediate(types.PollInterval, hc.healthCheckTimeout, func() (bool, error) {
```

Comment:
I think we just need the timeout part here, and do not need the retry, because the health check func is blocking IIRC. Is there a dedicated timeout util function? If we want to just utilize `PollImmediate` for the timeout logic, we can set the interval to be same as timeout value.

Reply:
Did not see a dedicated timeout function. Using the PollImmediate with timeout = pollinterval.